[CI] SearchableSnapshotRecoveryStateIntegrationTests testRecoveryStateRecoveredBytesMatchPhysicalCacheState failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/w3y6276h3es4g/tests/:x-pack:plugin:searchable-snapshots:internalClusterTest/org.elasticsearch.xpack.searchablesnapshots.recovery.SearchableSnapshotRecoveryStateIntegrationTests/testRecoveryStateRecoveredBytesMatchPhysicalCacheState

**Reproduction line:**
```
./gradlew ':x-pack:plugin:searchable-snapshots:internalClusterTest' --tests "org.elasticsearch.xpack.searchablesnapshots.recovery.SearchableSnapshotRecoveryStateIntegrationTests.testRecoveryStateRecoveredBytesMatchPhysicalCacheState" -Dtests.seed=37CF1F75CEB7D168 -Dtests.locale=el-CY -Dtests.timezone=America/Indiana/Petersburg -Druntime.java=20
```

**Applicable branches:**
8.8

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.xpack.searchablesnapshots.recovery.SearchableSnapshotRecoveryStateIntegrationTests&tests.test=testRecoveryStateRecoveredBytesMatchPhysicalCacheState

**Failure excerpt:**
```
java.lang.AssertionError: Physical cache size doesn't match with recovery state data
Expected: <9744L>
     but: was <10768L>

  at __randomizedtesting.SeedInfo.seed([37CF1F75CEB7D168:792DC5E477402C60]:0)
  at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
  at org.junit.Assert.assertThat(Assert.java:956)
  at org.elasticsearch.xpack.searchablesnapshots.recovery.SearchableSnapshotRecoveryStateIntegrationTests.testRecoveryStateRecoveredBytesMatchPhysicalCacheState(SearchableSnapshotRecoveryStateIntegrationTests.java:106)
  at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
  at java.lang.reflect.Method.invoke(Method.java:578)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1758)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:946)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:982)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:996)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:48)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:45)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:843)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:490)
  at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:955)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:840)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:891)
  at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:902)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:38)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at org.apache.lucene.tests.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
  at org.apache.lucene.tests.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:43)
  at org.apache.lucene.tests.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:44)
  at org.apache.lucene.tests.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:60)
  at org.apache.lucene.tests.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:47)
  at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:390)
  at com.carrotsearch.randomizedtesting.ThreadLeakControl.lambda$forkTimeoutingTask$0(ThreadLeakControl.java:850)
  at java.lang.Thread.run(Thread.java:1623)

```

Another one at https://gradle-enterprise.elastic.co/s/pau6gphccegsq/tests/:x-pack:plugin:searchable-snapshots:internalClusterTest/org.elasticsearch.xpack.searchablesnapshots.recovery.SearchableSnapshotRecoveryStateIntegrationTests/testRecoveryStateRecoveredBytesMatchPhysicalCacheState?top-execution=1

```
java.lang.AssertionError: Physical cache size doesn't match with recovery state data |  
-- | --
  | Expected: <17925L> |  
  | but: was <18859L> |  

...

REPRODUCE WITH: ./gradlew ':x-pack:plugin:searchable-snapshots:internalClusterTest' --tests "org.elasticsearch.xpack.searchablesnapshots.recovery.SearchableSnapshotRecoveryStateIntegrationTests.testRecoveryStateRecoveredBytesMatchPhysicalCacheState" -Dtests.seed=1C63BBE242E3FCBC -Dtests.locale=sr-ME -Dtests.timezone=Pacific/Fiji -Druntime.java=20

```

I am muting on main and 8.8. PR upcoming that you will see linked. @tlrx please consider assigning this ticket to you if your PR fixes this, and unmuting the test in main and 8.8.x, and referencing the ticket in your PR to close this.

--- a/docs/changelog/97278.yaml
+++ b/docs/changelog/97278.yaml
@@ -1,0 +1,6 @@
+pr: 97278
+summary: Fix reused/recovered bytes for files that are recovered from cache
+area: Snapshot/Restore
+type: bug
+issues:
+ - 95994

--- a/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/SearchableSnapshotDirectory.java
+++ b/x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/store/SearchableSnapshotDirectory.java
@@ -74,7 +74,6 @@ import java.util.Objects;
 import java.util.Set;
 import java.util.concurrent.Executor;
 import java.util.concurrent.atomic.AtomicBoolean;
-import java.util.concurrent.atomic.AtomicLong;
 import java.util.function.LongSupplier;
 import java.util.function.Supplier;
 
@@ -504,13 +503,9 @@ public class SearchableSnapshotDirectory extends BaseDirectory {
                     final IndexInput input = openInput(file.physicalName(), CachedBlobContainerIndexInput.CACHE_WARMING_CONTEXT);
                     assert input instanceof CachedBlobContainerIndexInput : "expected cached index input but got " + input.getClass();
 
-                    final AtomicLong prefetchedBytes = new AtomicLong(0L);
+                    final AtomicBoolean alreadyCached = new AtomicBoolean();
                     try (var fileListener = new RefCountingListener(ActionListener.runBefore(completionListener.acquire().map(v -> {
-                        // we don't support files to be reported as partially recovered from disk and partially from the blob store, but
-                        // this is something that can happen for fully mounted searchable snapshots. It is possible that prewarming
-                        // prefetched nothing if a concurrent search was executing (and cached the data) or if the data were fetched from
-                        // the blob cache system index.
-                        if (prefetchedBytes.get() == 0L) {
+                        if (alreadyCached.get()) {
                             recoveryState.markIndexFileAsReused(file.physicalName());
                         } else {
                             recoveryState.getIndex().addRecoveredFromSnapshotBytesToFile(file.physicalName(), file.length());
@@ -520,6 +515,7 @@ public class SearchableSnapshotDirectory extends BaseDirectory {
 
                         if (input instanceof CachedBlobContainerIndexInput cachedIndexInput) {
                             if (cachedIndexInput.getPersistentCacheInitialLength() == file.length()) {
+                                alreadyCached.set(true);
                                 logger.trace(
                                     () -> format(
                                         "%s file [%s] is already available in cache (%d bytes)",
@@ -540,7 +536,6 @@ public class SearchableSnapshotDirectory extends BaseDirectory {
                                     final long startTimeInNanos = statsCurrentTimeNanosSupplier.getAsLong();
                                     var prefetchedPartBytes = ((CachedBlobContainerIndexInput) input).prefetchPart(part, cancelPreWarming);
                                     if (prefetchedPartBytes > -1L) {
-                                        prefetchedBytes.addAndGet(prefetchedPartBytes);
                                         logger.trace(
                                             () -> format(
                                                 "%s part [%s/%s] of [%s] warmed in [%s] ms (%d bytes)",
@@ -549,7 +544,7 @@ public class SearchableSnapshotDirectory extends BaseDirectory {
                                                 file.numberOfParts(),
                                                 fileName,
                                                 timeValueNanos(statsCurrentTimeNanosSupplier.getAsLong() - startTimeInNanos).millis(),
-                                                prefetchedBytes
+                                                prefetchedPartBytes
                                             )
                                         );
                                     }

